### PR TITLE
Query refactoring

### PR DESCRIPTION
**Pull request by nnethercote**
Just some cleanups I found when learning about the query system.

Best reviewed one commit at a time.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -468,6 +468,18 @@ pub struct GlobalCtxt<'tcx> {
     pub(crate) alloc_map: Lock<interpret::AllocMap<'tcx>>,
 }
 
+impl<'tcx> GlobalCtxt<'tcx> {
+    /// Installs `self` in a `TyCtxt` and `ImplicitCtxt` for the duration of
+    /// `f`.
+    pub fn enter<'a: 'tcx, F, R>(&'a self, f: F) -> R
```
Comment (Zoxc) [inline compiler/rustc_middle/src/ty/context.rs:474]: This shouldn't be available outside `rustc_interface` as it cannot be accidentally called.

Comment (nnethercote) [reply compiler/rustc_middle/src/ty/context.rs:474]: Even before this PR, it's used in several other places: rustdoc, miri, etc. This `grep` output gives an idea:
```
compiler/rustc_driver_impl/src/lib.rs:                    queries.global_ctxt()?.enter(|tcx| {
compiler/rustc_driver_impl/src/lib.rs:            queries.global_ctxt()?.enter(|tcx| tcx.output_filenames(()));
compiler/rustc_driver_impl/src/lib.rs:            queries.global_ctxt()?.enter(|tcx| {
compiler/rustc_interface/src/queries.rs:            self.global_ctxt()?.enter(|tcx| {
compiler/rustc_interface/src/queries.rs:        let (crate_hash, prepare_outputs, dep_graph) = self.global_ctxt()?.enter(|tcx| {
src/librustdoc/doctest.rs:                let collector = queries.global_ctxt()?.enter(|tcx| {
src/tools/miri/src/bin/miri.rs:        queries.global_ctxt().unwrap().enter(|tcx| {
tests/run-make-fulldeps/obtain-borrowck/driver.rs:        queries.global_ctxt().unwrap().enter(|tcx| {
compiler/rustc_interface/src/queries.rs:            gcx.enter(|tcx| {
compiler/rustc_interface/src/queries.rs:                gcx.enter(rustc_query_impl::alloc_self_profile_query_strings);
compiler/rustc_interface/src/queries.rs:                .time("serialize_dep_graph", || gcx.enter(rustc_incremental::save_dep_graph));
src/librustdoc/lib.rs:            gcx.enter(|tcx| {
```

Comment (Zoxc) [reply compiler/rustc_middle/src/ty/context.rs:474]: I meant to say the upstream compiler crates cannot call it. The downstream crates however must call it to initialize the `GlobalCtxt` pointer in the TLS state, before getting access to `TyCtxt`. The `QueryContext` structure ensures they do so.

**Pull request merged**